ocaml/xapi/xapi_psr.ml

```diff
@@ -248,78 +348,114 @@ functor
       SecretString.write_to_file old_pool_secret_backup_path old_pool_secret ;
       SecretString.write_to_file new_pool_secret_backup_path new_pool_secret
 
-    let retrieve () =
-      match !Xapi_globs.pool_secrets with
-      | x :: y :: _ ->
-          (x, y)
-      | _ ->
-          (* do the backups exist? *)
-          raise
-            Api_errors.(
-              Server_error
-                (internal_error, ["can't retrieve backup pool secrets"]))
+    let retrieve = read_backups
 
-    let tell_accept_new_pool_secret (_, new_pool_secret) host =
+    let tell_accept_new_pool_secret (old_pool_secret, new_pool_secret) host =
       Helpers.call_api_functions ~__context (fun rpc session_id ->
           Client.Host.notify_accept_new_pool_secret rpc session_id host
-            new_pool_secret)
+            old_pool_secret new_pool_secret)
 
-    let tell_send_new_pool_secret (_, new_pool_secret) host =
+    let tell_send_new_pool_secret (old_pool_secret, new_pool_secret) host =
       Helpers.call_api_functions ~__context (fun rpc session_id ->
           Client.Host.notify_send_new_pool_secret rpc session_id host
-            new_pool_secret)
+            old_pool_secret new_pool_secret)
 
-    let tell_cleanup_old_pool_secret host =
+    let tell_cleanup_old_pool_secret (old_pool_secret, new_pool_secret) host =
       Helpers.call_api_functions ~__context (fun rpc session_id ->
-          Client.Host.cleanup_pool_secret rpc session_id host)
+          Client.Host.cleanup_pool_secret rpc session_id host old_pool_secret
+            new_pool_secret)
 
-    let cleanup_master =
-      cleanup_internal
-        ~additional_files_to_remove:[checkpoint_path; checkpoint_tmp_path]
+    let cleanup_master (old_ps, new_ps) =
+      cleanup_internal ~additional_files_to_remove:[checkpoint_path] ~old_ps
+        ~new_ps
   end
 
-let notify_new ~new_ps =
+let notify_new ~__context ~old_ps ~new_ps =
+  Assert.no_checkpoint () ;
   match !Xapi_globs.pool_secrets with
-  | _ :: _ :: _ ->
-      D.info "xapi_psr.ml:notify_new: already accepting new pool secret"
-  | _ ->
-      let old_ps = Xapi_globs.pool_secret () in
+  | [priority_1_ps; priority_2_ps] ->
+      (* check disk state *)
+      Assert.backups_match ~old_ps ~new_ps ;
+      (* check runtime state *)
+      if SecretString.(equal priority_2_ps new_ps && equal priority_1_ps old_ps)
```

Comment:
Has this been run through `make format`?

Reply:
It has